pdns/dnsdistdist/dnsdist-tcp-downstream.cc

```diff
@@ -505,8 +535,10 @@ void TCPConnectionToBackend::handleIO(std::shared_ptr<TCPConnectionToBackend>& c
         conn->d_ioState->update(iostate, handleIOCallback, conn, ttd);
       }
     }
+    skipped = conn->d_handleSkipped;
+    conn->d_handleSkipped = false;
   }
-  while (reconnected);
+  while (reconnected || skipped);
```

Comment:
I wonder if it wouldn't make sense to loop until we have nothing left to send or receive (or the connection has died, of course), something like:

```suggestion
  while (reconnected || (iostate != IOState::Done && !conn->d_connectionDied && !conn->d_lastIOBlocked));
```
we would have to add the `d_lastIOBlocked` boolean and update it when  `tryRead` or `tryWrite` returned anything other than `Done`.

Reply:
That doesn't directly work, because (as I wrote it currently) iostate is Done after the new query is received.

When I changed that to `IOState::NeedRead` in the `handleResponse` return, it works in the regression tests, but then unit tests fail again on the expected steps.

the (little dumb) `skipped` thing is only way I could make both integration tests and unit test go through :) but it doesn't "feel" correct.

Reply:
I completely understand. Do you think it would be possible to write a reproducer for the initial issue, perhaps by tweaking the Python code we use in our regression tests? It would be wonderful to be able to reproduce the issue outside of macOS.

Reply:
I will try to replicate on FreeBSD